pkg/virt-controller/services/launchercommandrenderer.go

```diff
@@ -0,0 +1,154 @@
+	launcherTimeout  int
+	ovmfPath         string
+	virtShareDir     string
+}
+
+type VirtLauncherCommandRendererOption func(renderer *VirtLauncherCommandRenderer)
+
+func NewVirtLauncherCommandRenderer(
+	vmUID string,
+	domainName string,
+	gracePeriodSeconds int,
+	hookSidecarListLen int,
+	vmNamespace string,
+	virtLauncherStaticConfig VirtLauncherStaticConfig,
+	opts ...VirtLauncherCommandRendererOption,
+) *VirtLauncherCommandRenderer {
+
+	return applyOptionsToRenderer(
+		&VirtLauncherCommandRenderer{
+			domainName:               domainName,
+			gracePeriodSeconds:       gracePeriodSeconds,
+			numberOfHookSidecars:     hookSidecarListLen,
+			namespace:                vmNamespace,
+			virtLauncherStaticConfig: virtLauncherStaticConfig,
+			vmUID:                    vmUID,
+		},
+		opts,
+	)
+}
+
+func applyOptionsToRenderer(virtLauncherCmdRenderer *VirtLauncherCommandRenderer, opts []VirtLauncherCommandRendererOption) *VirtLauncherCommandRenderer {
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		opt(virtLauncherCmdRenderer)
+	}
+	return virtLauncherCmdRenderer
+}
+
+func NewDoppelgangerPodRender(opts ...VirtLauncherCommandRendererOption) *VirtLauncherCommandRenderer {
+	return applyOptionsToRenderer(
+		&VirtLauncherCommandRenderer{isDoppelgangerPod: true},
+		opts,
+	)
+}
+
+func (vlcr *VirtLauncherCommandRenderer) Render() []string {
+	if vlcr.isDoppelgangerPod {
+		if vlcr.areAnyOptionsSet() {
+			log.Log.Warning("doppelganger pod does not allow options")
+		}
+		return []string{"/bin/bash",
+			"-c",
+			"echo", "bound PVCs"}
+	}
+
+	const launcherExecBinary = "/usr/bin/virt-launcher-monitor"
+	command := []string{
+		launcherExecBinary,
+		"--qemu-timeout", generateQemuTimeoutWithJitter(vlcr.virtLauncherStaticConfig.launcherTimeout),
+		"--name", vlcr.domainName,
+		"--uid", vlcr.vmUID,
+		"--namespace", vlcr.namespace,
+		"--kubevirt-share-dir", vlcr.virtLauncherStaticConfig.virtShareDir,
+		"--ephemeral-disk-dir", vlcr.virtLauncherStaticConfig.ephemeralDiskDir,
+		"--container-disk-dir", vlcr.virtLauncherStaticConfig.containerDiskDir,
+		"--grace-period-seconds", strconv.Itoa(vlcr.gracePeriodSeconds),
+		"--hook-sidecars", strconv.Itoa(vlcr.numberOfHookSidecars),
+		"--ovmf-path", vlcr.virtLauncherStaticConfig.ovmfPath,
+	}
+	if vlcr.isNonRoot {
+		command = append(command, "--run-as-nonroot")
+	}
+
+	if vlcr.customDebugFilters != "" {
+		command = append(command, "--libvirt-log-filters", vlcr.customDebugFilters)
+	}
+
+	if vlcr.allowEmulation {
+		command = append(command, "--allow-emulation")
+	}
+	if vlcr.keepAfterFailure {
+		command = append(command, "--keep-after-failure")
+	}
+	if vlcr.shouldSimulateCrash {
+		command = append(command, "--simulate-crash")
+	}
+	return command
+}
+
+func (vlcr *VirtLauncherCommandRenderer) areAnyOptionsSet() bool {
+	return vlcr.isNonRoot ||
+		vlcr.allowEmulation ||
+		vlcr.keepAfterFailure ||
+		vlcr.shouldSimulateCrash ||
+		vlcr.customDebugFilters != ""
+}
```

Comment:
This means that every fields that's added to `VirtLauncherCommandRenderer` needs to be added to this function.

How about simply checking whether `equality.Semantic.DeepEqual(vlcr, &VirtLauncherCommandRenderer{isDoppelgangerPod: true})`?

Reply:
Also, maybe you can rename the function to something like `isDoppelgangerPodRendererEmpty()`?

Reply:
Good idea.

Reply:
Done.

Is this name OK ?

Reply:
Awesome!